Fclfp hardik

hardikc07 opened the pull request.

Commit:
new sample - feature collection layer sample

Commit:
Update build.gradle

changes in build.gradle file

Commit:
Update settings.gradle

Commit:
changes in files for sample

Commit:
changes in files

Commit:
screenshot added and changes in java files

Commit:
Update README.md

JonLavi commented on feature_layers/feature-collection-layer-from-portal/README.md:
```diff
@@ -0,0 +1,34 @@
+# Feature Collection Layer From Portal
+
+This sample demonstrates opening a feature collection saved as a portal item
```
We're trying to keep the readme as close to the common desing as possible, so I'd suggest sticking with the same short description as we have there:

```suggestion
Create a feature collection layer from a portal item.
```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/README.md:
```diff
@@ -0,0 +1,34 @@
+## How to use the sample
+
+The feature collection is loaded from the Portal item when the sample starts. Enter feature collection saved as portal item in the text field
+and click on button to see result.
```
I don't think we need to manually break the lines here, as there's no official character limit per line.
I think we can also rephrase the instruction a tiny bit, because what we're really entering is the ID of the portal item. How about:
```suggestion
The feature collection is loaded from the Portal item when the sample starts. Enter the ID of a Portal item that holds a feature collection into the text field, and click on the button to see result.
```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/README.md:
```diff
@@ -0,0 +1,34 @@
+
+## How it works
+
+To display a `FeatureCollection` as a `portal Item` on an `ArcGISMap` :
```
We've decided in the readme review process to gemerally not start these sections with a sentence such as 'To do X...', so we could just remove this.
```suggestion
```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/README.metadata.json:
```diff
@@ -0,0 +1,35 @@
+{
+    "category": "Feature layers",
+    "description": "Combine feature tables with different geometries into a single layer.",
```
This looks like it's from another sample 😄 . It should be the same intro sentence as in the common design
```suggestion
    "description": "Create a feature collection layer from a portal item.",
```

hardikc07 replied:
that is true it was copied from another sample and forgot to update

JonLavi commented on feature_layers/feature-collection-layer-from-portal/README.metadata.json:
```diff
@@ -0,0 +1,35 @@
+    "description": "Combine feature tables with different geometries into a single layer.",
+    "ignore": false,
+    "images": [
+        "FeatureCollectionLayer.png"
```
This should also refer to the screenshot from this sample:
```suggestion
        "FeatureCollectionLayerFromPortal.png"
```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/README.metadata.json:
```diff
@@ -0,0 +1,35 @@
+        "SimpleFillSymbol",
+        "SimpleLineSymbol",
+        "SimpleMarkerSymbol",
+        "SimpleRenderer"
```
These should be the 'Tags' from the readme:
```suggestion
        "collection", 
        "feature collection", 
        "feature collection layer", 
        "id", 
        "item", 
        "map "notes, 
        "portal"
```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/README.metadata.json:
```diff
@@ -0,0 +1,35 @@
+        "SimpleMarkerSymbol",
+        "SimpleRenderer"
+    ],
+    "redirect_from": "/java/latest/sample-code/feature-collection-layer.htm",
```
I _think_ we leave this blank for new samples, and it gets automatically populated when the developer webpage is generated. I'll double check!

tschie replied:
This should only be here if this sample is meant to replace the feature collection layer sample.

JonLavi commented on feature_layers/feature-collection-layer-from-portal/README.md:
```diff
@@ -0,0 +1,34 @@
+* Feature
+* Portal
+* PortalItem
+
```
The readme is missing the 'about the data' section and the 'tags' from the common design:

```suggestion

## About the data

The sample uses a sample layer depicting [world populations](https://www.arcgis.com/home/item.html?id=32798dfad17942858d5eef82ee802f0b).

## Tags

collection, feature collection, feature collection layer, id, item, map notes, portal

```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/README.metadata.json:
```diff
@@ -0,0 +1,35 @@
+        "SimpleFillSymbol",
+        "SimpleLineSymbol",
+        "SimpleMarkerSymbol",
+        "SimpleRenderer"
```
This should match with the 'Relevant API section of the readme:
```suggestion
        "FeatureCollection"
        "FeatureCollectionLayer"
        "Portal"
        "PortalItem"
```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/README.metadata.json:
```diff
@@ -0,0 +1,35 @@
+        "SimpleRenderer"
+    ],
+    "snippets": [
+        "src/main/java/com/esri/samples/feature_collection_layer/FeatureCollectionLayerSample.java"
```
These should be the paths for all the files that are part of the sample code, from the `src` folder:
```suggestion
        "src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java"
```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/README.metadata.json:
```diff
@@ -0,0 +1,35 @@
+    "snippets": [
+        "src/main/java/com/esri/samples/feature_collection_layer/FeatureCollectionLayerSample.java"
+    ],
+    "title": "Feature Collection Layer"
```
And of course the sample's title 😄 
```suggestion
    "title": "Feature Collection Layer From Portal"
```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/README.md:
```diff
@@ -0,0 +1,34 @@
+
+This sample demonstrates opening a feature collection saved as a portal item
+
+![](FeatureCollectionLayerFromPortal.png)
```
We've been trying to populate the 'alt' text for the screenshot so it's consistent and accessible:

```suggestion
![Image of create feature collection layer portal item](FeatureCollectionLayerFromPortal.png)
```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalController.java:
```diff
@@ -0,0 +1,109 @@
+ */
+package com.esri.samples.feature_collection_layer_from_portal;
+
+import com.esri.arcgisruntime.data.FeatureCollection;
```
Generally the order we'd like the imports in is:
```
import java.*
<blank line>
import javafx.*
<blank line>
import com.esri.arcgisruntime.*
```

You can set up intellij to do this for you by going to 
settings > editor> code style> java > imports tab
Then edit the section 'import layout'. The UI is kind of terrible but you'll get there!

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalController.java:
```diff
@@ -0,0 +1,109 @@
+     * Handles searching the provided Item ID.
+     */
+    @FXML
+    private void fetchFromPortal() {
```
Nice work on figuring out FXML and the Controller setup!
I do think that the UI here is quite small, since it's just a text field and one button. So we can easily put it all just in the `FeatureCollectionLayerFromPortalSample.java' file and do everything from there. 
Check out the samples 'Change Feature Layer Renderer' and 'OpenMapURL', they are quite simple and somewhat similar. I think you could do it in a similar way, just build a small UI in the main Sample class. Give me a shout if you'd like any advice!

hardikc07 replied:
as a best practice can we still keep MVC pattern here for this sample ?

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalController.java:
```diff
@@ -0,0 +1,109 @@
+
+                // load the portal and add the portal item
+                Portal portal = new Portal("https://www.arcgis.com/");
+                System.out.println(FeatureCollectionItemIdTextField.getText());
```
Don't forget to get rid of all the console logging when you're finished 😉 Happens to me too, so no worries!

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalController.java:
```diff
@@ -0,0 +1,109 @@
+                // create feature collection and add to the map as a layer
+                FeatureCollection featureCollection = new FeatureCollection(portalItem);
+                FeatureCollectionLayer featureCollectionLayer = new FeatureCollectionLayer(featureCollection);
+                map.getOperationalLayers().add(featureCollectionLayer);
```
In general the pattern we try and promote is to only add the layers to the map's operational layers *after* we've checked that the appropriate resource has loaded successfuly.
So in this case I'd suggest putting this block of code where you create the FeatureCollection and Layer from the Portal inside the `if (portalItem.getLoadStatus() == LoadStatus.LOADED) {` statement on line 80

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2016 Esri.
```
Don't forget to update the copyright year 😉 

```suggestion
 * Copyright 2020 Esri.
```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,64 @@
+  @Override
+  public void start(Stage stage) throws IOException {
+    // set up the scene
+    FXMLLoader loader = new FXMLLoader(getClass().getResource("/fclfp.fxml"));
```
I don't think we need to keep the MVC structure here, but just pointing out that it would be best to name the `fxml` file with a more clear file name (generally the samples name with underscores), so `feature_collection_layer_from_portal.fxml`

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/resources/style.css:
```diff
@@ -0,0 +1,27 @@
+.panel-region .label {
```
Going forward we're trying to limit this CSS file to only the rules that are being used. Once you've changed the sample to only only live in one file, just check which of these you are using, and delete the rest.

JonLavi reviewed (changes requested):
Nice work @hardikc07! Well done on doing the full MVC pattern on this sample. It certainly is a good exercise. However I think in this sample we can make the UI in the main file, and don't need a `Controller` file and `fxml`, since it's a very small amount of UI (1 Text field and 1 button). Other than that, the code looks quite good. If you could restructure the sample to only be in one file, I'll have another closer look. Also, don't forget to update the README.metadata.json file. It's not the most exciting bit of work, but we definitely need it for the Developer Page to work correctly.

Commit:
Update feature_layers/feature-collection-layer-from-portal/README.md

Co-Authored-By: Jonathan Lavi <jonathan.lavi@gmail.com>

Commit:
Update feature_layers/feature-collection-layer-from-portal/README.md

Co-Authored-By: Jonathan Lavi <jonathan.lavi@gmail.com>

Commit:
Apply suggestions from code review

Co-Authored-By: Jonathan Lavi <jonathan.lavi@gmail.com>

Commit:
sample changes

removed controller file , added code into one java file and also removed fxml/css file

JonLavi commented on feature_layers/feature-collection-layer-from-portal/README.md:
```diff
@@ -0,0 +1,30 @@
+* FeatureCollectionLayer
+* Feature
+* Portal
+* PortalItem
```
We're still missing the 'About the data' and 'Tags' sections here:
```suggestion
* PortalItem

## About the data

The sample uses a sample layer depicting [world populations](https://www.arcgis.com/home/item.html?id=32798dfad17942858d5eef82ee802f0b).

## Tags

collection, feature collection, feature collection layer, id, item, map notes, portal

```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/README.metadata.json:
```diff
@@ -0,0 +1,28 @@
+        "feature collection layer", 
+        "id", 
+        "item", 
+        "map "notes, 
```
```suggestion
        "map",
        "notes, 
```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,151 @@
+import javafx.scene.control.TextField;
+import javafx.application.Application;
+import javafx.scene.layout.StackPane;
+import javafx.stage.Stage;
```
Let's make sure we order the imports in a way that it's consistent with the other samples.
You can also change some settings in intellij to have this done automatically for you.
```suggestion
import javafx.geometry.Insets;
import javafx.geometry.Pos;
import javafx.scene.Scene;
import javafx.scene.control.Alert;
import javafx.scene.control.Button;
import javafx.scene.control.TextField;
import javafx.application.Application;
import javafx.scene.layout.StackPane;
import javafx.stage.Stage;

import com.esri.arcgisruntime.data.FeatureCollection;
import com.esri.arcgisruntime.layers.FeatureCollectionLayer;
import com.esri.arcgisruntime.mapping.ArcGISMap;
import com.esri.arcgisruntime.mapping.Basemap;
import com.esri.arcgisruntime.mapping.view.MapView;
import com.esri.arcgisruntime.portal.Portal;
import com.esri.arcgisruntime.portal.PortalItem;
```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,151 @@
+import javafx.scene.layout.StackPane;
+import javafx.stage.Stage;
+
+
```
There's an extra empty line here we don't need.
```suggestion
```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/README.metadata.json:
```diff
@@ -0,0 +1,28 @@
+        "FeatureCollection"
+        "FeatureCollectionLayer"
+        "Portal"
+        "PortalItem"
```
These are missing some commas behind them for the json to parse properly:
```suggestion
        "FeatureCollection",
        "FeatureCollectionLayer",
        "Portal",
        "PortalItem"
```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,151 @@
+
+    @Override
+    public void start(Stage stage) {
+        // set up the scene
```
We should probably rephrase this so that users don't confuse this scene with an ArcGIS Scene: 
```suggestion
        // set up the application scene
```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,151 @@
+            map = new ArcGISMap(Basemap.createOceans());
+            // create a view for this ArcGISMap and set ArcGISMap to it
+            mapView = new MapView();
+            mapView.setMap(map);
```
The comments here don't quite line up with the order of what the code is doing.
I've noticed other samples do this as well. I'll create an issue for us to tidy these up a little and make them more uniform.

For this sample, I'd suggest rearranging a bit:
```suggestion

            // create a new ArcGISMap with an oceans Basemap
            map = new ArcGISMap(Basemap.createOceans());
            
            // create a map view and set the ArcGISMap to it
            mapView = new MapView();
            mapView.setMap(map);
```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,151 @@
+    private Portal portal;
+    private FeatureCollectionLayer featureCollectionLayer;
+    private FeatureCollection featureCollection;
+    private TextField input;
```
In the samples repo we generally try and give all variables a name that implies their type. So perhaps we can name this: `TextField inputTextField`

hardikc07 replied:
i have changed textfiled name to inputTextField

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,151 @@
+            input.setText("32798dfad17942858d5eef82ee802f0b");
+
+            // create button to perform action
+            Button fetchFromPortal = new Button("Open from portal item");
```
As with the text field, it's probably best for the user's understanding to call this `fetchFromProtalButton`.

hardikc07 replied:
changed button name to fetchFromProtalButton

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,151 @@
+
+            /**
+             * Action event on button and validation of empty Text Filed
+             */
```
We generally only do javadoc style comments (those with `/**`) on methods or classes.
Since this is just a block of code you can do:
```suggestion
             // verify the input and fetch the portal item
```

JonLavi replied:
The comment still appears in JavaDoc style for me. Could you just double check it's in a single line?

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,151 @@
+                if (!"".equals(input.getText())) {
+                    fetchFeatureCollectionFromPortal();
+                } else {
+                    new Alert(Alert.AlertType.ERROR, "Portal Itemid is empty. Please enter a portal item id.").show();
```
We could make this prompt a bit more uniform in terms of formatting
```suggestion
                    new Alert(Alert.AlertType.ERROR, "Portal Item ID is empty. Please enter a Portal Item ID.").show();
```

JonLavi replied:
We can do this check for the contents of the `inputTextBox` directly in the method `fetchFeatureCollectionFromPortal()`. That way all you have to do is:
```
fetchFromPortal.setOnAction(e -> fetchFeatureCollectionFromPortal());
```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,151 @@
+            mapView = new MapView();
+            mapView.setMap(map);
+
+            // create text field to input user's own coordinate string
```
Let's update this to specify it's a portal item ID that's being taken
```suggestion
            // create text field to input user's own portal item ID
```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,151 @@
+
+            });
+
+            fetchFeatureCollectionFromPortal();
```
I'd suggest moving this to line 72 (directly after creating the field input). That way It's clear what the input is for that you specify at the beginning of the sample.

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,151 @@
+            fetchFeatureCollectionFromPortal();
+
+            // add feature layer to ArcGISMap
+            map.getOperationalLayers().add(featureCollectionLayer);
```
I'd suggest putting this inside the method `fetchFeatureCollectionFromPortal()`, after the line `featureCollectionLayer = new FeatureCollectionLayer(featureCollection);`. That way it's clear to the user that they have to add their new layer to the map in order for it to show up.

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    /**
+     * Fetch feature collection from portal and load as feature layer
```
In JavaDoc-style comments, it's common to end sentences with a period:
```suggestion
     * Fetch feature collection from portal and load as feature layer.
```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,151 @@
+     */
+    private void fetchFeatureCollectionFromPortal() {
+
+        //create portal and portal item
```
Missing a space here before the comment:
```suggestion
        // create portal and portal item
```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,151 @@
+    private void fetchFeatureCollectionFromPortal() {
+
+        //create portal and portal item
+        portal = new Portal("https://www.arcgis.com/");
```
I think we can move this line into the Start method, because it's always going to be a copy of the same Portal object. So if we have it in the 'start' method, it's only ever being created once, and re-used every time when creating the `PortalItem` in the next line.

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,151 @@
+        // create feature collection and add to the map as a layer
+        featureCollection = new FeatureCollection(portalItem);
+        featureCollectionLayer = new FeatureCollectionLayer(featureCollection);
+
```
This is where we'd add the FCL to the map. Then it would also fit the comment 👍

tschie replied:
Yep, you can also make `featureCollectionLayer` a local variable instead of a field. You'll need to remember to clear the map's operational layers before adding a new one.

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,151 @@
+        featureCollection = new FeatureCollection(portalItem);
+        featureCollectionLayer = new FeatureCollectionLayer(featureCollection);
+
+        featureCollectionLayer.loadAsync();
```
If you add the line `map.getOperationalLayers().add(featureCollectionLayer);` above, it would automatically trigger loading the layer, so you can get rid of the 'loadAsync();` call here.
I think generally the trick with loadables is that if you're adding them to the map or performing some sort of operation with them, they will start loading by themselves. If you just instantiate them and do nothing else, you'd need the `loadAsync()`.

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,151 @@
+     * Fetch feature collection from portal and load as feature layer
+     */
+    private void fetchFeatureCollectionFromPortal() {
+
```
Let's do the check `!input.getText().isEmpty()` in this method.

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,151 @@
+
+        //create portal and portal item
+        portal = new Portal("https://www.arcgis.com/");
+        PortalItem portalItem = new PortalItem(portal, input.getText());
```
You can do the check for the contents of the portal item directly after this, then continue creating the FCL. At the end of the block you can then show an error if it's not a valid item:
``` 
if (portalItem.getType() == PortalItem.Type.FEATURE_COLLECTION) {
  // create a feature collection from the portal item
  // create a feature collection layer
  // add the feature collection layer to the map
else {
  // show the error
}
```

JonLavi commented on gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,5 +1,6 @@
+#Wed Mar 25 17:43:43 PDT 2020
```
I don't think we need to keep this comment in here.

JonLavi reviewed (commented):
Nice work @hardikc07 ! It's looking a lot more concise now. There's just a bit of rearranging of the code I would suggest, so that it's easier for users to follow the main workflow. I've also found a few places where we can make the comment style more consistent. Give me a shout if you have any questions!

Commit:
Apply suggestions from code review

making changes as per review by jonathan

Co-Authored-By: Jonathan Lavi <jonathan.lavi@gmail.com>

tschie commented on feature_layers/feature-collection-layer-from-portal/README.metadata.json:
```diff
@@ -0,0 +1,29 @@
+        "id", 
+        "item", 
+        "map",
+        "notes, 
```
missing `"` which is why GitHub is showing this as invalid JSON.
```suggestion
        "notes", 
```

tschie commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,150 @@
+            stage.show();
+
+
+            // create a new ArcGISMap with an oceans Basemap
```
```suggestion
            // create a new ArcGISMap with the oceans basemap
```

tschie commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,150 @@
+            // create a new ArcGISMap with an oceans Basemap
+            map = new ArcGISMap(Basemap.createOceans());
+            
+            // create a map view and set the ArcGISMap to it
```
```suggestion
            // create a map view and set the map to it
```

tschie commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,150 @@
+            // create text field to input user's own portal item ID
+            input = new TextField();
+            input.setMaxWidth(250);
+            input.setText("32798dfad17942858d5eef82ee802f0b");
```
can set the default text with the constructor

tschie commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,150 @@
+
+             // verify the input and fetch the portal item
+            fetchFromPortal.setOnAction(e -> {
+
```
no need for these extra lines

tschie commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,150 @@
+
+        featureCollectionLayer.loadAsync();
+        featureCollectionLayer.addDoneLoadingListener(() -> {
+            if (portalItem.getType() != PortalItem.Type.FEATURE_COLLECTION) {
```
Should do check that item successfully loaded first. There should also be an else condition which shows an alert with a generic error message.

tschie reviewed (changes requested):
A few changes needed. I agree with Jon's feedback too. Good work so far!

Commit:
changes in 2 files

made changes as per suggestion in pr

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -16,6 +16,7 @@
 
 package com.esri.samples.feature_collection_layer_from_portal;
 
+import com.esri.arcgisruntime.loadable.LoadStatus;
```
Just make sure to put this import in the right order (with the bloc below).
I can really recommend you set up IntelliJ to order the imports for you, so you won't ever have to worry about it again.

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -58,45 +58,38 @@ public void start(Stage stage) {
+
+            // create a map view and set the map to it
             mapView = new MapView();
             mapView.setMap(map);
```
I'd suggest swapping the order of these two lines. That way we deal with the map and map view first, and then create the portal. It should be a bit easier to follow.

```suggestion
            // create a map view and set the map to it
            mapView = new MapView();
            mapView.setMap(map);
            
            // create portal
            portal = new Portal("https://www.arcgis.com/");
```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -58,45 +58,38 @@ public void start(Stage stage) {
-            });
+            inputTextField = new TextField();
+            inputTextField.setMaxWidth(250);
+            inputTextField.setText("32798dfad17942858d5eef82ee802f0b");
```
You could do as Tyler suggested and put the contents of the text field in the constructor:
```suggestion
            inputTextField = new TextField("32798dfad17942858d5eef82ee802f0b");
            inputTextField.setMaxWidth(250);
```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -58,45 +58,38 @@ public void start(Stage stage) {
+            inputTextField.setMaxWidth(250);
+            inputTextField.setText("32798dfad17942858d5eef82ee802f0b");
 
             fetchFeatureCollectionFromPortal();
```
Let's add a small comment here to help the users see what's going on:

```suggestion
			// fetch the default feature collection layer from the portal
            fetchFeatureCollectionFromPortal();
```

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -108,25 +101,34 @@ public void start(Stage stage) {
+            new Alert(Alert.AlertType.ERROR, "Portal Item ID is empty. Please enter a Portal Item ID.").show();
+        }
 
     }
```
Nice work on this! There's only a couple of things left to tidy here:
A bit of whitespace would help make the users read this a little easier. As a general rule of thumb, we try and add a blank line between each block of code.

For the sake of making an easy to follow sample, we'd want to do a few things and consider the order of doing them:
1. check that the portal item ID is provided
2. check that we can load the portal item, and that it loaded successfuly (Adding the FCL should only happen after we check that the portal item has done loading successfuly. That way we don't perform the action before the appropriate resource is loaded. )
3. check that the portal item is a feature collection (We need to check that the portal item is a feature collection before actually trying to use it to create the FCL )
4. create the feature collection layer
5. add the later to the map.

This would make it a lot easier for users to read the sample and understand in which order stuff needs to be done.

Keep in mind that we don't generally print out the loading error using `portalItem.getLoadError().getCause().toString()`, as you've done, because it will be displayed to they end-user, and they wouldn't know what to do with the technical information. Instead we just write a message that describes what went wrong.

```suggestion
    /**
     * Fetch a feature collection from a portal and load it as a feature layer.
     */
    private void fetchFeatureCollectionFromPortal() {
        if (!inputTextField.getText().isEmpty()) {
            // crate portal item
            PortalItem portalItem = new PortalItem(portal, inputTextField.getText());

            // load the portal item
            portalItem.loadAsync();
            portalItem.addDoneLoadingListener(() -> {
                if (portalItem.getLoadStatus() == LoadStatus.LOADED) {

                    // check that the portal item is a feature collection
                    if (portalItem.getType() == PortalItem.Type.FEATURE_COLLECTION) {

                        // create a feature collection layer
                        featureCollection = new FeatureCollection(portalItem);
                        FeatureCollectionLayer featureCollectionLayer = new FeatureCollectionLayer(featureCollection);

                        // add the feature layer to the map
                        map.getOperationalLayers().clear();
                        map.getOperationalLayers().add(featureCollectionLayer);

                    } else {
                        new Alert(Alert.AlertType.ERROR, "Portal Item is not a valid Feature Collection.").show();
                    }
                } else {
                    new Alert(Alert.AlertType.ERROR, "Failed to load Portal Item.").show();
                }
            });
        } else {
            new Alert(Alert.AlertType.ERROR, "Portal Item ID is empty. Please enter a Portal Item ID.").show();
        }
    }
    
```

JonLavi reviewed (commented):
This is almost ready @hardikc07, nice work! I'm just suggesting a bit of reordering to the method that creates the FCL, so that it's easier to follow for users. Let me know what you think!

Commit:
made changes in sample

few more changes as per review by jonathan
- removed javadoc comment
- set text in costructor
- import order / map portal oder fixes
- validation of portalitem and loading status before creating FC

hardikc07 reviewed (commented):
made changes as suggested

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -61,27 +61,25 @@ public void start(Stage stage) {
             fetchFeatureCollectionFromPortal();
 
             // create button to perform action
             Button fetchFromProtalButton = new Button("Open from portal item");
```
There's a typo in the name of the button:
`fetchFromPortalButton`

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -101,29 +99,36 @@ public void start(Stage stage) {
+            });
+
+        } else
         {
```
We don't need the newline here between `else` and `{`

JonLavi commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -101,29 +99,36 @@ public void start(Stage stage) {
-                        }
+            portalItem.loadAsync();
+
+            portalItem.addDoneLoadingListener(()->{
```
Just a bit of formatting here:
```suggestion
            portalItem.addDoneLoadingListener(() -> {
```

JonLavi reviewed (approved):
Nice work @hardikc07 ! 
Only found a typo and a little bit of formatting to fix, but after that I think it's good.

Commit:
minor changes

- typo in button name
- remove extra sapce for else cond
- added space in listener method

tschie commented on feature_layers/feature-collection-layer-from-portal/README.md:
```diff
@@ -0,0 +1,38 @@
+3. Verify that the item represents a feature collection.
+4. Create a `FeatureCollection` from the item.
+5. Create a `FeatureCollectionLayer`, referring to the feature collection.
+6. Add the feature collection layer to the map's `OperationalLayers` collection.
```
```suggestion
6. Add the feature collection layer to the map's operational layers collection.
```

tschie commented on feature_layers/feature-collection-layer-from-portal/README.metadata.json:
```diff
@@ -0,0 +1,28 @@
+    "snippets": [
+        "src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java"
+    ],
+    "title": "Feature Collection Layer From Portal"
```
```suggestion
    "title": "Feature Collection Layer from Portal"
```

tschie commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,155 @@
+                        new Alert(Alert.AlertType.ERROR, "This is not valid Feature Collection.").show();
+                    }
+                } else {
+                    new Alert(Alert.AlertType.ERROR, "Item does not exist or is inaccessible").show();
```
```suggestion
                    new Alert(Alert.AlertType.ERROR, "Portal item failed to load").show();
```

tschie commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,155 @@
+                        map.getOperationalLayers().add(featureCollectionLayer);
+
+                        } else {
+                        new Alert(Alert.AlertType.ERROR, "This is not valid Feature Collection.").show();
```
```suggestion
                          new Alert(Alert.AlertType.ERROR, "Portal item is not a feature collection").show();
```

tschie commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,155 @@
+    private MapView mapView;
+    private ArcGISMap map;
+    private Portal portal;
+    private FeatureCollection featureCollection;
```
This should be a local variable in the fetch method.

tschie commented on feature_layers/feature-collection-layer-from-portal/src/main/java/com/esri/samples/feature_collection_layer_from_portal/FeatureCollectionLayerFromPortalSample.java:
```diff
@@ -0,0 +1,155 @@
+            });
+        } else
+        {
+            new Alert(Alert.AlertType.ERROR, "Portal Item ID is empty. Please enter a Portal Item ID.").show();
```
```suggestion
            new Alert(Alert.AlertType.ERROR, "Portal Item ID is empty").show();
```

tschie reviewed (approved):
Just a few more small things to change. Please mark the feedback as Resolved when you make the updates.

Commit:
done changes tyler suggested

- alert message
- readme
